core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -923,7 +974,39 @@ public InstallState getInstallState() {
      */
     @Restricted(NoExternalUse.class)
     public void setInstallState(@Nonnull InstallState newState) {
+        if(newState.equals(installState)) {
+            return; // no change, nothing to do
+        }
+        if(InstallState.INITIAL_SETUP_COMPLETED.equals(installState)) {
+            throw new IllegalArgumentException("Setup has already been completed!");
+        }
         installState = newState;
+        try {
+            switch(newState) {
+            case INITIAL_PLUGINS_INSTALLED:
+                setInstallState(InstallState.CONFIGURING_SECURITY);
```

Comment:
Why not set this state in the code that actually configures the security? Seems wrong to do it here as part of setting the state to `INITIAL_PLUGINS_INSTALLED`.